tests/baselines/reference/cannotIndexError.errors.txt

```diff
@@ -0,0 +1,14 @@
+cannotIndexError.ts(4,5): error TS2860: Type 'string' cannot be used to index type 'T', because type 'T' is generic and an index signature of a constraint of a generic type can only be used for reading.
+cannotIndexError.ts(4,5): error TS2860: Type 'symbol' cannot be used to index type 'T', because type 'T' is generic and an index signature of a constraint of a generic type can only be used for reading.
+
+
+==== cannotIndexError.ts (2 errors) ====
+    // From #47357
+    
+    function foo<T extends Record<string | symbol, any>>(target: T, p: string | symbol) {
+        target[p] = "";
+        ~~~~~~~~~
+!!! error TS2860: Type 'string' cannot be used to index type 'T', because type 'T' is generic and an index signature of a constraint of a generic type can only be used for reading.
+        ~~~~~~~~~
+!!! error TS2860: Type 'symbol' cannot be used to index type 'T', because type 'T' is generic and an index signature of a constraint of a generic type can only be used for reading.

```

Comment:
It seems bad that there are two errors reported.

I think you can drop the words "of a constraint" and simplify to "index signature of a generic type".

Reply:
I thought briefly about making only one error be reported, but it might be a bit wonky given the structure of the existing code...
We individually check each element of the union type used for indexing (e.g. in `T[string | symbol]` we check both `string` and `symbol`), and while for this error it makes sense to treat `string | symbol` as a single thing and report only one error, it's not always the case and I don't know how we could tell which approach to take beforehand, e.g. in this example: https://www.typescriptlang.org/play?ts=5.3.0-dev.20230928#code/GYVwdgxgLglg9mABMOcBMAeAKogpgDylzABMBnRMEAWwCNcAnAbQF1EAyRAb0SbIC5EZKAxhgA5i0FU6jRAB8hIseMQBfAHwAKKAEMG43FEFYANIgAOg4aIkLKNegwCU3AFCJEeg0aYW2ALyIAIwA3G5qQA

Reply:
Gotcha. There is already some logic where we try to handle a union of `boolean` instead of `true` and `false` individually in `getIndexedAccessTypeOrUndefined` to avoid 2 errors for what it's worth, so I think there's an alternative path where you might be able to collect the types that fail and then error on those instead.

Reply:
I'm going to suggest changing the error message to `Type '{0}' is generic and can only be indexed for reading`. I would just omit the index type, it adds no value. As a bonus, this means the error message is the same for `string` and `symbol`, so it should get reduced to a single occurrence in error reporting.

Reply:
I really like that message, but what about the example I linked to above, where `number` can be used to index into a generic type for writing, but `string`/`symbol` cannot?

Reply:
I'm not too concerned by that example. I mean, the example is trying to index with something that isn't permitted in a write position, so the error is correct for the given situation. It is true that there might not be an error for a subset of the index types, but that's not what the example is doing.